### PR TITLE
Rmd 22827 detailed report refactor

### PR DESCRIPTION
**Pull request by DavidGamba**
@tonytvo please review.

### CODE REVIEW

File: src/main/groovy/com/calgaryscientific/gradle/VeracodeDetailedReport.groovy [hunk 1]
```diff
@@ -27,10 +27,46 @@
...
+     * Component tree:
+     *
+     * /component @description @file_name @library @max_cvss_score @sha1 @vendor @version @vulnerabilities
+     *  /file_paths
+     *      /file_path @value
+     *  /vulnerabilities
+     *      /vulnerability @cve_id @cve_summary @cvss_score @cwe_id @severity
+     *  /violated_policy_rules
+     */
+    static List<List<String>> softwareCompositionAnalysisRows(Node xml) {
```
Comment (tonytvo) [inline src/main/groovy/com/calgaryscientific/gradle/VeracodeDetailedReport.groovy:54]: ok .... ok ... static function here ....... :) :)

File: src/main/groovy/com/calgaryscientific/gradle/VeracodeDetailedReport.groovy [hunk 2]
```diff
@@ -27,10 +27,46 @@
...
+        List<List<String>> rows = []
+        List<String> componentFields = ['library', 'file_name', 'vendor', 'description']
+        List<String> vulnerabilityFields = ['cve_id', 'cwe_id', 'cvss_score', 'severity', 'cve_summary']
+        // header row
+        rows.add(componentFields + vulnerabilityFields)
+        for (Node component : XMLIO.getNodeList(xml, 'software_composition_analysis', 'vulnerable_components', 'component')) {
```
Comment (tonytvo) [inline src/main/groovy/com/calgaryscientific/gradle/VeracodeDetailedReport.groovy]: not sure if you would be interested in this groovy idiom
```
componentWithVulnerabilities = XMLIO.getNodeList(xml, 'software_composition_analysis', 'vulnerable_components', 'component').filter { it -> it.(component.attribute('vulnerabilities') as Integer) > 0 }
rowVulnerabilities = componentWithVulnerabilities.collect {  component -> XMLIO.getNodeList(component, 'vulnerabilities', 'vulnerability')).collect { [component, it]} }
result = rowVulnerabilities.collect { component, Vulnerability -> MLIO.getNodeAttributes(component, componentFields) + XMLIO.getNodeAttributes(vulnerability, vulnerabilityFields) }
```

you might need to flatten the list.....

**Review by tonytvo — APPROVED** [review]
other than concerning static functions :) :) , I think it looks good

File: src/main/groovy/com/calgaryscientific/gradle/VeracodeDetailedReport.groovy (hunk 2 above)
Comment (DavidGamba) [reply src/main/groovy/com/calgaryscientific/gradle/VeracodeDetailedReport.groovy]: Thanks Tony, I refactored it with the `.each` idiom to be consistent with the rest of the codebase. Other than that, I don't think adding the filter and the extra lines adds to clarity.

**Pull request merged**